File: epi_judge_java/epi/LowestCommonAncestor.java
package epi;

import epi.test_framework.*;

public class LowestCommonAncestor {
  public static BinaryTreeNode<Integer> lca(BinaryTreeNode<Integer> tree,
                                              BinaryTreeNode<Integer> node0,
                                              BinaryTreeNode<Integer> node1) {
        // Implement this placeholder.
        // algo: do pre-order (left -> right -> root)
        // gather info about right and left children
        // need to return the first node in which has 1 node in it's left subtree and one node in it's right subtree.
        // if 1 node is not found in each subtree,
        // return left subtree count + right subtree count + 1 (if the root itself contains the data).

        // botton line after testing
        // all good except I ignored a case where the the same node0 and node1 are the same node.
        // to me 18 minutes to write
        LCAResult lcaRes = calcLCA(tree, node0, node1);
        return lcaRes.lca;
    }

    static LCAResult calcLCA(BinaryTreeNode<Integer> tree,
                             BinaryTreeNode<Integer> node0,
                             BinaryTreeNode<Integer> node1) {
        if (tree == null) {
            return new LCAResult(null, 0);
        }
        LCAResult leftRes = calcLCA(tree.left, node0, node1);
        if (leftRes.lca != null) {
            return leftRes;
        }
        LCAResult rightRes = calcLCA(tree.right, node0, node1);
        if (rightRes.lca != null) {
            return rightRes;
        }
        int foundInRoot = 0;
        if (tree == node0) {
            foundInRoot++;
        }
        if (tree == node1) {
            foundInRoot++;
        }
        if ((leftRes.nodesFound + rightRes.nodesFound + foundInRoot) == 2) {
            return new LCAResult(tree, 2);
        } else {
            return new LCAResult(null, leftRes.nodesFound + rightRes.nodesFound + foundInRoot);
        }
    }

    static class LCAResult {
        BinaryTreeNode lca;
        int nodesFound;

        LCAResult(BinaryTreeNode lca, int nodesFound) {
            this.lca = lca;
            this.nodesFound = nodesFound;
        }
    }


    @EpiTest(testDataFile = "lowest_common_ancestor.tsv")
    public static int lcaWrapper(TimedExecutor executor,
                                 BinaryTreeNode<Integer> tree, Integer key0,
                                 Integer key1) throws Exception {
        BinaryTreeNode<Integer> node0 = BinaryTreeUtils.mustFindNode(tree, key0);
        BinaryTreeNode<Integer> node1 = BinaryTreeUtils.mustFindNode(tree, key1);

        BinaryTreeNode<Integer> result =
        executor.run(() -> lca(tree, node0, node1));

        if (result == null) {
            throw new TestFailure("Result can not be null");
        }
        return result.data;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "LowestCommonAncestor.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_python/lowest_common_ancestor.py
# Also available on leetcode
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/

import functools
from typing import Optional, NamedTuple

from binary_tree_node import BinaryTreeNode
from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node, strip_parent_link
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


# 09:44 - start writing after reading the general idea of the solution
# if the LCA is in the current path of the parent (meaning not in both left and right subtrees)
#   then the LCA will be set

class LcaRes(NamedTuple):
    found_so_far: int  # number of nodes in the current path
    lca: Optional[BinaryTreeNode]  # if lca not found yet then None

# this is a more complex solution that will terminate as soon we found the 2 children
def lca(tree: BinaryTreeNode, node0: BinaryTreeNode,
        node1: BinaryTreeNode) -> Optional[BinaryTreeNode]:
    def calc_lca(curr: BinaryTreeNode,
                 found_so_far: int  # number of LCAs found so far
                 ) -> LcaRes:
        if not curr:
            return LcaRes(found_so_far, None)
        found_in_curr = [node0, node1].count(curr)
        if found_in_curr == 2:
            if found_in_curr == 2:
                return LcaRes(2, curr)

        found_so_far += found_in_curr
        if found_so_far == 2:
            return LcaRes(2, None)

        left_res = calc_lca(curr.left, found_so_far)
        if left_res.lca:
            return left_res
        if (left_res.found_so_far > found_so_far):
            found_in_curr += 1
        if found_in_curr == 2:
            return LcaRes(2, curr)

        found_so_far = left_res.found_so_far

        right_res = calc_lca(curr.right, found_so_far)
        if right_res.lca:
            return right_res

        if right_res.found_so_far > found_so_far:
            found_in_curr += 1

        found_so_far = right_res.found_so_far
        if found_in_curr == 2:
            return LcaRes(2, curr)
        else:
            return LcaRes(found_so_far, None)

    # todo handle edge cases that the roor is the LCA
    if not tree:
        return None

    return calc_lca(tree, 0).lca


@enable_executor_hook
def lca_wrapper(executor, tree, key1, key2):
    strip_parent_link(tree)
    result = executor.run(
        functools.partial(lca, tree, must_find_node(tree, key1),
                          must_find_node(tree, key2)))

    if result is None:
        raise TestFailure('Result can\'t be None')
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('lowest_common_ancestor.py',
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))
